Add option max_levels to decode

Kuroneer opened the pull request:
This PR adds the option `max_levels` to `jiffy:decode`.
This option allows you to provide a `{max_levels, non_neg_integer()}` as option. Jiffy then decodes un to N levels for the JSON, and the ones deeper than that are returned as `{json, binary()}` instead.

This PR couples nicely with #139, allowing to `jiffy:encode(jiffy:decode(Binary, [{max_levels, X}]))`

Why is it useful?
In my case, I use erlang+jiffy to route streamed jsons to different endpoints based on a routing key that's near the top of the document. Decoding and encoding the whole document every time seems to be wasteful.
A simple test with this shows a great improvement for me, however, I don't know if it'd be worth to include for everyone:
```
2> timer:tc(fun() -> [jiffy:decode(JSON, [return_maps]) || _ <- lists:seq(1, 10000)], ok end).
{2355549,ok}
3> timer:tc(fun() -> [jiffy:decode(JSON, [return_maps, {max_levels, 1}]) || _ <- lists:seq(1, 10000)], ok end).
{564704,ok}
```

Kuroneer committed:
Support max_levels option in decoder

Kuroneer committed:
Max_levels tests

davisp commented:
I'm not immediately in love with this. My first thought is wether [jsx][https://github.com/talentdeficit/jsx] would be a better fit with its SAX style parsing? Do you need to modify the JSON as you route it? I've been reluctant to merge a PR like #139 because it means that Jiffy's output isn't guaranteed to be correct since its not checking the pre-encoded data.

Kuroneer commented:
Thanks for pointing me to jsx, I didn't know about that project. However, if I read it correctly, it would not suit my needs because unfortunately, yes, I need to modify the root node, I'd need to test the performance too, because this PR avoids creating binaries and such that are not going to exist in the end result.

I understand your concerns, would you be more keen if I included some way to validate `{json, binary()}` tuples to #139 (with an option to trust the binaries)? I guess that since it's a `json_value()`, it'd suffice to try to jiffy:decode the binary with max_levels to 0 (which would need to be allowed, right now it's a pos_integer()). Also, I'd need to make sure that all the checks are kept, as some of them are avoided if the term is not actually needed.

davisp commented:
Yeah, JSX being an option would have relied on that key being "early" in the JSON doc and not require mutation.

My biggest concern around partially encoded JSON is around Jiffy emitting malformed JSON. If we can come up with an implementation that is not overly complex and can be guaranteed to not produce invalid JSON then I'd be on board.

Previously when I've noodled over this I was expecting to almost have two parser implementations when I more or less rejected as an option since it doubles the amount of logic that needs testing. However, this PR makes me think there might be a decent approach to handling things such that normal uses of the API would be able to guarantee things enough for me (I'm ignoring cases of users that reverse engineer any internal state to intentionally side step guarantees, or in other words, it doesn't have to be iron clad, just not allow for accidental malformed JSON).

I hadn't really thought about attempting a no-op decoding approach like you did for your max levels. I think what leads to an interesting approach along the lines of:

```
jiffy:validate_json(binary(), Opts) -> true|false.
```

Implementing a validation feature requires that we tweak the decoder structure to run through *all* checks while not generating Erlang terms. Throwing a boolean on the Decoder struct and then gating all calls would be straightforward I believe. Note that this is different than your current approach since it bypasses some checks.

Once we have that, then I'd add a second function that was something like:

```
jiffy:prepare_partial_json(Binary, State, Options) -> partial_json().`
jiffy:extract_partial_json(partial_json()) -> {Binary, State}.
```

This would use the no-op decoding to ensure that the partial binary is in fact valid JSON. The state is required to know that we stop and end at the appropriate places. It'd be something along the lines of `value | kv_pairs | array_elems` atoms. The return value would be a resource object that contained the state information and a reference to the binary. This could then be placed within any Erlang JSON representation at an appropriate location.

When we then go through encoding we'd extend the partial encoding loop that currently only handles bignums to stitch the JSON together which will require us adding/checking comma separators and the like and making assertions on the partial json. I.e., that you can't stick KV pairs inside an array and generate something like `[1.0, "foo": "bar", false]` or something weird.

Then getting back to your original goal of having a max_levels decoding, we'd just flip on the partial encoding flag dynamically during decoding and return the opaque references that can then be fed back into the encoder without a second pass (which I am assuming covers your ask for disabling re-checking already validated JSON).

And then after all of that we write a whole bunch of tests.

davisp commented:
One thing I'd note is that this approach does not allow something funky that affects multiple levels from multiple places. I.e., having two places where partial JSON is inserted to create valid JSON after encoding. I.e., having something like:

```erlang
[
    {json, <<"{\"foo\": {\"bar\":">>},
    true,
    {json, <<"}}">>}
]
```

Or something. But I can't even figure out what that would look like.

davisp commented:
Though, it occurs to me that we could add an encoder option that would encode values and return `partial_json()` resources and then you could do something like:

```
Partial = jiffy:encode(Term, [partial]),
jiffy:encode({[{<<"foo">>, {[{<<"bar">>, Partial}]}]}).
```

Kuroneer commented:
I'll work on it and give it a try

Kuroneer committed:
Full json validation with max_levels option

Kuroneer committed:
Add partial option to encode

Kuroneer commented:
Although I'm not completely sold on resources due to the serialization limitation and the need for a process-independent environment to hold the binary, it works nicely.

Can you check the PR and give me your thoughts?

Kuroneer commented:
(sorry for the late reply, it's been a really hectic week)

Just to note the changes, in 89ac61b I've updated the validator to check everything. The only thorn left is building the binary for a string if it has_escapes to make use of the constructed binary, I'll see how it ends playing with the offset and the buffer. (Objects are ok by construction)

In dc1fd56 the resource is created and returned, currently only for json_values().

I'll work next on having a validate(), (which in my head is only a alias for {max_levels,0})

Kuroneer committed:
Manage copy_strings with max_levels option again

Kuroneer committed:
Add validate/1,2 call

Kuroneer committed:
Add validate/1,2 to README.md

Kuroneer committed:
Avoid binary creation for strings if unneeded

Kuroneer commented:
`jiffy:validate/1,2` ready.

davisp commented:
Looks like you're going in the right direction. I'll try and find time to give this a thorough review.

Kuroneer committed:
Refine types

Kuroneer committed:
Make jiffy:json_raw() public

jiffy:json_raw is a reference(), so its opaqueness is stored within it.
With this change, it allows to check if a subterm is a json_raw with
is_reference/1

nickva commented:
Looking at clearing some old issues and PRs before a 2.0 release, I can't say I am excited about this PR, it adds quite a bit more complexity. I could maybe see an extra API call to just validate without decoding or maybe if there is a relatively small change we can accomplishing the same but as it stand the size of the PR and the complexity is too much I think so I'll close the PR.

Kuroneer commented:
Its purpose wasn't as much as validation as avoiding creating terms
without purpose. Thus not needing to allocate memory, or freeing it.

In our use case (an API gateway), this change alone amounted to a ~5% CPU
reduction, although this was before JIT was introduced; I'd need to revisit
this today to see if it's worth it now.

We've been using this PR for the past years without issue, tbh, I had lost
hope of seeing this merged.

In any case, just keep it in mind if you find the same issue in the future,
as it can save quite a bit of $ in the right circumstances.

nickva commented:
@Kuroneer my main issue with this PR was how it's implemented: it changing too much code for what it attempts to do. Partial encoding of `{json, IoData}` I didn't mind adding as it was using a neat trick from @davisp and it was just a two lines of code effectively. So it's not that I am against the idea in general, but maybe there can be a simpler implementation. I am thinking something like a single function which knows how to skip over json, it would do the matching of string starts/ends, {, }, [, ] and bump a single depth integer up and down.
